src/AppInstallerRepositoryCore/Microsoft/Schema/1_4/DependenciesTable.cpp

```diff
@@ -0,0 +1,375 @@
+    {
+        SQLite::Savepoint savepoint = SQLite::Savepoint::Create(connection, std::string{ s_DependenciesTable_Table_Name } + "update_dependencies_v1_4");
+
+        const auto dependencies = GetDependencies(manifest, Manifest::DependencyType::Package);
+        auto existingDependencies = GetDependenciesByManifestRowId(connection, manifestRowId);
+
+        // Get dependencies to add.
+        std::set<Manifest::Dependency> toAddDependencies;
+        std::copy_if(
+            dependencies.begin(),
+            dependencies.end(),
+            std::inserter(toAddDependencies, toAddDependencies.begin()),
+            [&](Manifest::Dependency dep)
+            {
+                auto depRow = GetPackageRowIdAndVersion(connection, dep);
+                return existingDependencies.find(depRow.value()) == existingDependencies.end();
+            }
+        );
+
+        // Get dependencies to remove.
+        std::vector<SQLite::rowid_t> toRemoveDependencies;
+        std::vector<Manifest::Dependency> missingPackageNodes;
+        std::set<std::pair<SQLite::rowid_t, Utility::NormalizedString>> dependenciesPackageIdAndVersion;
+        std::for_each(
+            dependencies.begin(),
+            dependencies.end(),
+            [&](auto& dep)
+            {
+                auto depRow = GetPackageRowIdAndVersion(connection, dep);
+                if (!depRow.has_value())
+                {
+                    missingPackageNodes.emplace_back(dep);
+                    return;
+                }
+
+                dependenciesPackageIdAndVersion.emplace(depRow.value());
+            }
+        );
+
+        ThrowOnMissingPackageNodes(missingPackageNodes);
+
+        std::for_each(
+            existingDependencies.begin(),
+            existingDependencies.end(),
+            [&](std::pair<std::pair<SQLite::rowid_t, Utility::NormalizedString>, SQLite::rowid_t> row)
+            {
+                if (dependenciesPackageIdAndVersion.find(row.first) == dependenciesPackageIdAndVersion.end())
+                {
+                    toRemoveDependencies.emplace_back(row.second);
+                }
+            }
+        );
+
+        InsertManifestDependencies(connection, manifestRowId, toAddDependencies);
+        RemoveDependenciesByRowIds(connection, toRemoveDependencies);
+        savepoint.Commit();
+
+        return true;
+    }
+
+    void DependenciesTable::RemoveDependencies(SQLite::Connection& connection, SQLite::rowid_t manifestRowId)
+    {
+        SQLite::Savepoint savepoint = SQLite::Savepoint::Create(connection, std::string{ s_DependenciesTable_Table_Name } + "remove_dependencies_by_manifest_v1_4");
+        SQLite::Builder::StatementBuilder builder;
+        builder.DeleteFrom(s_DependenciesTable_Table_Name).Where(s_DependenciesTable_Manifest_Column_Name).Equals(manifestRowId);
+
+        builder.Prepare(connection).Execute(true);
+        savepoint.Commit();
```

Comment:
>Commit [](http://example.com/codeflow?start=18&length=6)

Since you are also the potential owner of the version ids, you need to collect them and remove if they are no longer needed.

This also brings up the point that you need to update `OneToOneTableDeleteIfNotNeededById` to handle the fact that you are now another potential reference point.

I would suggest changing the code there slightly to allow a more generic call out to each of the potential owners to check for reference.  That way dependencies can simply ignore any non-version calls as no reference.

Reply:
Can you explain:
**_I would suggest changing the code there slightly to allow a more generic call out to each of the potential owners to check for reference. That way dependencies can simply ignore any non-version calls as no reference._**

Reply:
In the function [`OneToOneTableDeleteIfNotNeededById`](https://github.com/microsoft/winget-cli/blob/2e693852fe7fae7f1643fdb799653645cdf0034f/src/AppInstallerRepositoryCore/Microsoft/Schema/1_0/OneToOneTable.cpp#L145), the only thing that is checked right now is whether the manifest table references the given value or not.

You will also need a similar "does the dependencies table reference the version value" check in that function.  But rather than adding that directly, I'm suggesting a change to that function to call new method like `ManifestTable::IsValueReferenced` and then also call `DependenciesTable::IsValueReferenced` so that they can do the check themselves.  Dependencies would not even check the database unless the value was a version or package id.  That sets a better pattern for extension than putting special logic to handle the dependencies directly into the other function.